### ISSUE TITLE
Possible bug with `ActiveRecord::AttributeMethods#attributes` and new columns

### ISSUE THREAD

**Issue by BrentWheeldon** [rails/rails#33793]
### Steps to reproduce

1. Launch a rails console
2. In another process run a migration to add a column to an existing table
3. Back in the console, inspect the output of `ActiveRecord::AttributeMethods#attributes`

[Script](https://gist.github.com/BrentWheeldon/023cfb846d327e4c52d74b83c8fd1f9b)

### Expected behavior
It does not include the new column (since including the new column makes `add_column` non-zero-downtime-safe)

### Actual behavior
The new column is included

### System configuration
**Rails version**: master

**Ruby version**: 2.5.1p57

### More context

I'm not sure if this is desired behaviour, or a bug. I'd be happy to work on a PR to fix this if folks agree that there is an issue here.

We noticed this issue during a deploy which added a new column to an existing table. We migrate the DB and still have all of our processes running the old version of the code and then do a rolling restart on our processes once the migrations are complete. As soon as the new column was added we started seeing errors because we have code that essentially does `ModelClass.new(model_instance.attributes.except(:id, :foo))` to duplicate records.

**Comment by y-yagi** [comment]
How about using `ignored_columns`? https://github.com/rails/rails/pull/21720

**Comment by rafaelfranca** [comment]
It is not possible given the way rails works.

To avoid doing database calls to discovery the columns to the database before every SQL we cache the result in the process. It means that while the process is running, if any database column is added or removed the cache will have outdated information. To force the cache to be regenerated you can run `Model.reset_column_information`.

**Comment by BrentWheeldon** [comment]
@rafaelfranca I think you’ve misunderstood the issue. The problem is that `#attributes` doesn’t use the cached column information - it returns attributes that the model doesn’t know about causing the failure in the test case in the gist.

**Comment by joniles** [comment]
We've hit the same issue (Rails 4.2.10, Ruby 2.4.3). In our case with have code similar to this:
```ruby
my_new_record = MyModel.new(my_old_record.attributes.except(:id))
```
If a new column is added to the `my_new_models` table after Rails has started, `attributes` will include an entry for that column causing the assignment to fail.

A workaround for us is to introduce a call to `compact`. The new column will be `nil` in `attributes` and any existing attributes with a `nil` value come from nullable columns, so we can just strip out the `nil` values entirely.

```ruby
my_new_record = MyModel.new(my_old_record.attributes.compact.except(:id))
```
This is useful where we know we are likely to encounter this issue, but a fix which means that `attributes` ignores columns added after Rails starts would be great.

**Comment by rafaelfranca** [comment]
Ahhh. Yeah, that is expected behavior too. This is why we added ignored_columns.

**Comment by BrentWheeldon** [comment]
I just opened a PR with a possible fix - #33811 - would love some feedback!
